lib/absinthe/phase/document/execution/resolution.ex

```diff
@@ -202,37 +202,74 @@ defmodule Absinthe.Phase.Document.Execution.Resolution do
     fun.(res, [])
   end
 
-  defp build_result(%{errors: []} = res, exec, _source, path) do
+  defp build_result(%{errors: errors} = res, exec, source, path) do
     %{
-      value: result,
+      value: value,
       definition: bp_field,
-      extensions: extensions
+      extensions: extensions,
     } = res
 
     full_type = Type.expand(bp_field.schema_node.type, exec.schema)
 
     bp_field = put_in(bp_field.schema_node.type, full_type)
 
-    result = result |> to_result(bp_field, full_type)
+    # if there are any errors, the value is always nil
+    value = case errors do
+      [] -> value
+      _ -> nil
+    end
+
+    errors = maybe_add_non_null_error(errors, value, full_type)
 
-    %{result | extensions: extensions}
+    value
+    |> to_result(bp_field, full_type, extensions)
+    |> add_errors(Enum.reverse(errors), &put_result_error_value(&1, &2, bp_field, source, path))
     |> walk_result(bp_field, full_type, exec, path)
+    |> propagate_null_trimming
```

Comment:
@bruce In the end I decided this was best accomplished within the Resolution phase itself. If we don't handle it here then the phase produces a non-spec compliant result tree, which each variant of the Result phase has to handle for itself. It's relatively important that the Result phases be kept simple because they're a key part of stuff like the Phoenix controller integration, any extension work, and so forth.

There's also a handy performance perk to doing it in this phase since we just make it part of the general stack unwinding process that builds the result tree, instead of doing a dedicated pass afterward.